### PR TITLE
Added setattr to `functional_call`.

### LINKED ISSUE #77133: Overriding `__getattribute__` is not enough in `stateless.functional_call`
### 🐛 Describe the bug

In `_stateless.functional_call`, it is not sufficient to override `__getattribute__` - we also need to override `__setattr__`.

```
from torch.nn.utils.stateless import functional_call
import torch.nn as nn
import torch
from functorch.compile import memory_efficient_fusion, print_compile, draw_graph_compile, default_decompositions

class Foo(nn.Module):
    def __init__(self):
        super().__init__()
        self.register_buffer('foo', torch.zeros(()))

    def forward(self, x):
        self.foo = self.foo + 1
        return x + self.foo

a = {'foo': torch.zeros(())}
mod = Foo()
print(mod.foo)
functional_call(mod, a, torch.ones(()))
print(mod.foo)
```

results in 
```
tensor(0.)
tensor(1.)
```

In other words, despite overriding a value with `functional_call`, we are still modifying the original buffer with `__setattr__`.

We probably need to override `__setattr__` to also look at the `_functional_values` dict.

cc: @albanD @jbschlosser @emcastillo 

### Versions

Nightly

cc @albanD @mruberry @jbschlosser @walterddr @kshitij12345

### PR DESCRIPTION
**Pull request by Chillee**
Fixes https://github.com/pytorch/pytorch/issues/77133

### CODE REVIEW

**Review by emcastillo — APPROVED** [review]
Great catch! I didn't thought of this case which is a clearly common one.
Could you add a minor test for this with the example you posted in the issue?

**Comment by Chillee** [comment]
@albanD wasn't super convinced this is the right thing to do, which is why I marked the PR WIP.

**Review by jbschlosser — COMMENTED** [review]
I see that the original `parameters_and_buffers` dict passed to `functional_call` is unmodified, and only the local copy of the dict stored on the module changes.

As it is now, users can't retrieve the post-call values and I could see it being useful to get those. wdyt @albanD ?

**Comment by Chillee** [comment]
I changed the code a decent chunk (cc: @emcastillo).

However, one thing that this `setattr` change has revealed is that the DataParallel test doesn't actually ... work, I think.

https://github.com/pytorch/pytorch/blob/master/torch/nn/parallel/replicate.py#L148 

The `nn.DataParallel` code pulls out params from `module._parameters` (which ignores the functional args!), and then setattrs them on the module. I don't totally understand what's going on, but considering that the `setattr` and access of `_parameters` both completely ignores the `functional_call` stuff, I suspect it's not doing the right thing.

**Comment by Chillee** [comment]
@albanD The current `stateless.functional_call` behavior does not work with `nn.DataParallel`

```
import torch
import torch.nn as nn
import torch.nn.utils.stateless as stateless

class Foo(nn.Module):
    def __init__(self):
        super().__init__()
        self.weight = nn.Parameter(torch.ones(5))

    def forward(self, x):
        return self.weight + x

mod = Foo().cuda()
mod = nn.DataParallel(mod, [0, 1])
print(stateless.functional_call(mod, {'module.weight': torch.zeros(5, device='cuda')}, (torch.ones(2, 5, device='cuda'),)))
```

errors with
```
RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cuda:1!
```

The only reason the current test passes is that it's being tested with scalar CUDA tensors, which can implicitly be coerced to any device freely.

Given that the existing `functional_call` doesn't work with `nn.DataParallel`, I'm gonna skip the test and land the PR.

File: test/test_stateless.py [hunk 1]
```diff
@@ -73,6 +73,7 @@ def test_functional_call_with_jit(self):
             self._run_call_with_mock_module(traced_module)
 
     @unittest.skipIf(not TEST_MULTIGPU, 'multi-GPU not supported')
+    @unittest.skip("This doesn't work right now")
```
Comment (albanD) [inline test/test_stateless.py:76]: Can you open an issue to track this please?

File: torch/nn/utils/stateless.py [hunk 2]
```diff
@@ -8,43 +8,51 @@
 
 # We avoid typing module here because module attributes are declared as Union[Parameter, Tensor] by default
 # and using other types causes mypy errors
-def _change_class(module) -> None:
+def _change_class(module, params_and_buffers) -> None:
     cls = module.__class__
-    func_params : Dict[str, Tensor] = module._functional_parameters
+    func_params : Dict[str, str] = module._attr_to_path
```
Comment (albanD) [inline torch/nn/utils/stateless.py:13]: nit variable name not updated?

**Review by albanD — COMMENTED** [review]
The doc needs updating and a small example there sounds like it would help to make things clear.

File: test/test_stateless.py (hunk 1 above)
Comment (Chillee) [reply test/test_stateless.py:76]: Done: https://github.com/pytorch/pytorch/issues/77576

**Comment by Chillee** [comment]
@albanD what's the doc you're referring to here? The docstring?

**Review by albanD — APPROVED** [review]
Thanks

**Comment by albanD** [comment]
Letting @jbschlosser take a look as well.

**Review by jbschlosser — APPROVED** [review]
LGTM, thanks for the updates :)

**Pull request closed without merging**